runtime/src/iree/hal/drivers/vulkan/vulkan_device.cc

```diff
@@ -1238,23 +1403,88 @@ static iree_status_t iree_hal_vulkan_device_query_i64(
-    } else if (iree_string_view_equal(
-                   key, iree_make_cstring_view("vulkan-spirv-fb-ptr"))) {
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("vulkan-spirv-fb-ptr"))) {
       // SPIR-V with device addresses is optionally supported based on whether
       // we have device feature support.
       *out_value = iree_all_bits_set(
                        device->logical_device->enabled_features(),
                        IREE_HAL_VULKAN_FEATURE_ENABLE_BUFFER_DEVICE_ADDRESSES)
                        ? 1
                        : 0;
+      return iree_ok_status();
+    }
+  }
+
+  // Note that the device queries used here should match the ones used in
+  // mapToDeviceQuery() on the compiler side.
+  if (iree_string_view_equal(category, IREE_SV("hal.dispatch"))) {
+    if (iree_string_view_equal(key, IREE_SV("compute.f16"))) {
+      bool v = device->logical_device->supported_properties().compute_f16;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("compute.f64"))) {
+      bool v = device->logical_device->supported_properties().compute_f64;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("compute.i8"))) {
+      bool v = device->logical_device->supported_properties().compute_i8;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("compute.i16"))) {
+      bool v = device->logical_device->supported_properties().compute_i16;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("compute.i64"))) {
+      bool v = device->logical_device->supported_properties().compute_i64;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("storage.8bit"))) {
+      bool v = device->logical_device->supported_properties().storage_8bit;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("storage.16bit"))) {
+      bool v = device->logical_device->supported_properties().storage_16bit;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("subgroup.arithmetic"))) {
+      uint32_t ops =
+          device->logical_device->supported_properties().subgroup_operations;
+      *out_value =
+          iree_all_bits_set(ops, VK_SUBGROUP_FEATURE_ARITHMETIC_BIT) ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("subgroup.shuffle"))) {
+      uint32_t ops =
+          device->logical_device->supported_properties().subgroup_operations;
+      *out_value =
+          iree_all_bits_set(ops, VK_SUBGROUP_FEATURE_SHUFFLE_BIT) ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("dotprod.4xi8.i32"))) {
```

Comment:
same here (and for all of these) - if there's some feature family with some number of options under 64 (because we return an i64) it's best to group them as coarsely as possible